Feat: 사용자 취향 설정 및 조회 API 구현

Kevin-jwd opened the pull request:
Closes #27 
## 개요

- #27 
- 선호 장르 설정/조회 기능을 위한 DTO 정의 및 Swagger 문서화, 커스텀 예외 처리 구현

## 작업사항

- **요청 DTO 정의**
  - `GenreIdsDto`: 장르 ID 배열 구조 정의 및 유효성 검사
  - `UpdatePreferenceRequestDto`: 영화/TV 시리즈 선호 장르를 함께 받는 구조
- **응답 DTO 정의**
  - `PreferenceDetailDto`: 사용자 선호 장르 목록 및 개수 응답
  - `GetPreferenceResponseDto`: 영화/TV 시리즈 선호 장르를 함께 반환하는 최상위 구조
- **커스텀 예외 처리**
  - 장르가 존재하지 않을 경우 커스텀 예외 `GenreNotFoundException` 처리

## 주의사항

- 현재 DTO 구조는 `movie`, `tv`에 한정되어 있으나, 추후 콘텐츠 유형이 확장될 경우 해당 구조 및 DTO 재설계가 필요할 수 있습니다.

Kevin-jwd committed:
Feat: 취향 관련 DTO 생성 및 Swagger 문서화 적용
- GenreIdsDto: 장르 ID 목록 검증
- UpdatePreferenceRequestDto: 영화/TV 장르 업데이트 요청
- GetPreferenceResponseDto: 영화/TV 선호 장르 응답 DTO
- PreferenceDetailDto: 선호 장르 상세 정보 DTO

Kevin-jwd committed:
Feat: 팩토리 메서드 추가

Kevin-jwd committed:
Feat: 장르 미발견 시 커스텀 예외 처리 추가

Kevin-jwd committed:
Feat: 사용자 취향 설정 및 조회 API 추가 및 로직 구현

Kevin-jwd committed:
Feat: ContentType 필드 추가

hightuv commented on src/preference/preference.service.ts:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  // User 검증
+  private async getUserOrThrow(userId: number): Promise<User> {
+    const user = await this.userRepository.findOneBy({ id: userId });
+    if (!user) throw new UserNotFoundException();
+    return user;
+  }
```
타입 가드 목적으로 Entity 조회 시에 항상 이런 식으로 작성하곤 했는데, 확실히 추출하니까 나은 것 같네요.
userId로 회원을 조회하는 거는 다른 모듈들에서도 사용하는 경우가 좀 있는데, 이 함수를 추후에 다른 곳에서도 쓸 수 있도록 하면 어떨까 싶기도 하네요.
추후에 UserService에서 이 로직을 추가해두고, UserService를 DI받아서 getUserOrThrow를 호출하는 식으로 하면 어떨까요?
조금 더 User와 관련된 곳에서 처리하는 게 맞을 듯 합니다.

Kevin-jwd replied:
함수로 분리해둘 생각만 했지 책임이 어디에 있는지 고려를 못했던 것 같네요
말씀대로 `getUserOrThrow`와 아래의 `getGenresOrThrow`는 각각의 도메인에 위치시키는 게 좋을 것 같습니다
본 브랜치에서 작업한 후 커밋하겠습니다

hightuv replied:
네 좋습니다. 그리고 id로 조회하는 경우가 있고, email로 조회하는 경우가 있는데, 이를 모두 포괄하는 함수면 좋지 않을까 싶네요!

hightuv commented on src/preference/preference.service.ts:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  // Genre 존재 검증
+  private async getGenresOrThrow(genreIds: number[]): Promise<Genre[]> {
+    const genres = await this.genreRepository.findBy({ id: In(genreIds) });
+    if (genres.length !== genreIds.length) throw new GenreNotFoundException();
+    return genres;
+  }
```
이 부분도 앞서 말씀드렸던 UserService처럼, GenreService에 구현해두고 DI받아서 사용하는 게 낫지 않을까 싶습니다.

hightuv commented on src/preference/preference.service.ts:
```diff
@@ -0,0 +1,115 @@
+    private readonly genreRepository: Repository<Genre>,
+  ) {}
+
+  async updateUserPreferences(
+    userId: number,
+    dto: UpdatePreferenceRequestDto,
+  ): Promise<void> {
+    const user = await this.getUserOrThrow(userId);
+    const genreIds = this.extractGenreIds(dto);
+    await this.preferenceRepository.delete({ user: { id: userId } });
+    if (genreIds.length === 0) return;
+
+    const genres = await this.getGenresOrThrow(genreIds);
+    const preferences = this.createPreferences(user, genres);
+    await this.preferenceRepository.save(preferences);
+  }
```
이 과정이 회원 취향 생성과 수정에서 모두 수행되는 로직인거죠?

hightuv reviewed (commented):
DTO쪽은 API 스펙에 맞게 잘 구현된 것으로 보입니다.
코멘트 확인해주시고, 의견 남겨주세요.

만약 getOrThrow를 해당되는 Service쪽에 구현하는 것에 동의하는 쪽이라면, 이 PR 이후에 따로 수정하고 적용하면 될 것 같습니다.
고생하셨습니다!

Kevin-jwd committed:
Feat: 사용자/장르 도메인 로직 분리 및 취향 업데이트 트랜잭션 적용

- PreferenceService에서 사용자/장르 조회 로직을 UserService/GenreService로 이동
- PreferenceService의 취향 업데이트 로직에 트랜잭션 추가

Kevin-jwd committed:
Feat: 영화/TV 시리즈 장르 조회 시 다중 ID 지원 및 TMDB API 호출 에러 처리 중앙화

- 장르 쿼리 파라미터에 ParseArrayPipe 적용
- 다중 장르 ID를 배열로 받아 TMDB API 쿼리 파라미터 구성 로직 확인
- tmdb.util.ts: fetchFromTmdb 함수에서 API 호출 실패 시 ExternalApiException 반환

hightuv commented on src/preference/preference.service.ts:
```diff
@@ -18,30 +18,34 @@ export class PreferenceService {
-    @InjectRepository(User)
-    private readonly userRepository: Repository<User>,
+    @InjectDataSource()
+    private readonly dataSource: DataSource,
```
DB가 하나만 연결된 경우에는 `@InjectDataSource()` 데코레이터 없이도 DI가 정상적으로 동작합니다.
여러 DB를 사용할 때만 데이터소스 이름을 명시적으로 지정하기 위해 데코레이터의 인자를 활용한다고 알고 있습니다.
저희는 DB 하나만 사용해서 지금은 이 데코레이터가 없어도 되기는 하지만, 있어도 상관없긴 합니다.
언급해두면 좋을 거 같아서 코멘트 남깁니다!

Kevin-jwd replied:
데코레이터 삭제해도 정상 동작하네요
몰랐던 내용을 알게 되었네요. 좋은 지적 감사합니다!

hightuv commented on src/user/user.service.ts:
```diff
@@ -95,4 +86,15 @@ export class UserService {
     return { isDuplicate: true, provider: findUser.provider };
   }
+
+  async getOrThrowById(id: number): Promise<User> {
+    const user = await this.userRepository.findOneById(id);
+    if (!user) throw new UserNotFoundException();
+    return user;
+  }
+  async getOrThrowByEmail(email: string): Promise<User> {
+    const user = await this.userRepository.findOneByEmail(email);
+    if (!user) throw new UserNotFoundException();
+    return user;
+  }
```
Entity를 조회하고, 예외처리를 하는 패턴은 다양한 곳에서 사용될 수 있을 거 같긴 합니다.
상황마다 다르게 메소드를 구현하기 보다는, 여러 인자를 받을 수 있게 하여 하나의 함수로 공통적으로 처리하는 방식을 고민해봐도 좋을 것 같습니다.
추후 리팩토링할 때 천천히 진행해도 좋아보입니다.

Kevin-jwd replied:
좋은 의견입니다
우선적으로 직관적이게 구현을 해두었는데 추후 통합 메서드로 관리하는 쪽이 좋아 보이네요

hightuv commented on src/movie/movie.controller.ts:
```diff
@@ -110,8 +110,11 @@ export class MovieController {
-    @Query('genreId') genreId: number,
-    @Query('page') page?: number,
+    @Query('genreId', new ParseArrayPipe({ items: String, optional: true }))
+    genreId: string[],
```
Query에 대한 DTO를 만들고, 해당 DTO에서 validator와 transformer로 유효성 검사와 변환을 처리해도 좋을 것 같습니다.

Kevin-jwd replied:
확실히 Query Parameter도 validator를 사용하여 검증을 하기 위해서는 DTO로 정의해두어야 좋을 것 같네요.
추가하겠습니다!

hightuv commented on src/tv/tv.controller.ts:
```diff
@@ -110,7 +110,8 @@ export class TvController {
   async getTvsByGenre(
-    @Query('genreId') genreId: number,
+    @Query('genreId', new ParseArrayPipe({ items: String, optional: true }))
+    genreId: string[],
```
MovieController에서의 코멘트 내용처럼 수정해도 좋을 것 같습니다.

hightuv reviewed (commented):
코멘트 확인해주시면 감사하겠습니다 :)

Kevin-jwd committed:
Feat: 쿼리 파라미터 DTO 통합 및 적용

hightuv reviewed (approved):
DTO로 변경한 거 확인했습니다.
추후 전반적으로 리팩토링이 필요할 것 같기는 한데, 다른 작업 진행에 필요한 부분이다보니 지금은 이대로 합치는 게 좋을 거 같습니다.
